server/routes/statistics.js

```diff
@@ -0,0 +1,226 @@
+	
+	console.log("Processing db data");
+}
+
+function sendBack(res, data) {
+	res.setHeader("Cache-Control", "no-store");
+	console.log("Sending back");
+	return res.status(200).send(data);
+}
+
+router.get("/byuser", async (req, res) => {
+	const name = req.query.username;
+	console.log(name);
+	
+	try {
+		if (requestCache.getStats().keys == 0) {
+			console.log("Refreshing cache");
+			await requestAndCache();
+		}
+		console.log("here");
+		if (!users.has(String(name))) {
+			console.log("Getting user info from database");
+			const userObj = await User.find({ username: name });
+			console.log(userObj)
+			sendBack(res, { data: userObj });
+		} else {
+			console.log("Getting user info from cache map");
+			const user = users.get(userId);
+			console.log("User")
+			res.setHeader("Cache-Control", "no-store");
+			sendBack(res, { data: user });
+		}		
+	} catch (error) {
+		
+	}
+})
+
+router.get("/standings", async (req, res) => {
+	try {
+		//If cache needs to be refreshed retrieve the list of users.
+		if (requestCache.getStats().keys == 0) {
+			console.log("Refreshing cache");
+			await requestAndCache();
+		}
+		
+		let pageSkip = req.query.page-1;
+		let sorting = req.query.sortBy;
+		let order = req.query.order;
+		console.log("fetching standings");
+		//Cache function caches the results of the standings if it has not been cached with that ordering already.
+		const cacheFunction = (sortFunc, sortByVal) => {
+			console.log("In caching function")
+			if (!sortBy.has(sortByVal)) {
+				console.log("Cache doesn't have it");
+				const allUsers = requestCache.get("allUsers");
+				const sortedUsers = [...allUsers];
+				console.log("Retrieved cache");
+				sortedUsers.sort(sortFunc);
+				sortBy.set(sortByVal, sortedUsers);
+			}
+			
+			if (order == "asc" && !sortBy.has(sortByVal+"Reverse")) {
+				const inOrder = sortBy.get(sortByVal);
+				const reverseOrder = [];
+				for (let i = inOrder.length-1;i>=0;i--) reverseOrder.push(inOrder[i]);
+				sortBy.set(sortByVal+"Reverse", reverseOrder);
+			}
+			
+			console.log("Returning value");
+			let values;
+			if (order == "asc") {
+				values = sortBy.get(sortByVal+"Reverse");
+			} else {
+				values = sortBy.get(sortByVal);
+			}
+			
+			let index = pageSkip * 50;
+			let lastIndex = index + 50;
+			if (index>=values.length) return res.status(200).send({ message: "None" });
+			if (lastIndex >= values.length) lastIndex = values.length;
+			let page = values.slice(index, lastIndex);
+			
+			return sendBack(res, { start: index+1, data: page });
+		}
+		//Cache by each sort criteria
+		if (sorting=="numWins") {
+			const sortFunction = function(x, y) {
+					const val1 = x.numWins;
+					const val2 = y.numWins;
+					return val2 - val1;
+				};
+			return cacheFunction(sortFunction, "numWins");
+		} else if (sorting=="numLosses") {
+			const sortFunction = function(x, y) {
+					const val1 = x.numLosses;
+					const val2 = y.numLosses;
+					return val2 - val1;
+				};

```

Comment:
There are lots of repeated code here.... for example you could use sorting as a variable and do x[sorting] and remove a lot of the repeated code here

Reply:
I was thinking of doing that actually.